File: auto_dealer/events/service_job_card.py
"""
Service Job Card Event Handlers
---------------------------------
Wired by hooks.py doc_events exactly as per PDF Section 4.1:

    "Service Job Card": {
        "on_submit": "auto_dealer.events.service_job_card.on_submit",
    },

On submit actions (PDF Section 9.1 + 9.2):
  - Check stock for required parts → auto-create Material Request if below required qty
  - WhatsApp notification to customer that job has started
"""

import frappe
from frappe import _
from frappe.utils import flt


def on_submit(doc, method=None):
    """
    On submit: check parts stock and auto-create Material Request if needed.
    Source: PDF Section 9.2.
    """
    _check_parts_and_create_material_request(doc)
    _send_job_started_whatsapp(doc)


def _check_parts_and_create_material_request(doc):
    """
    When a Service Job Card is submitted, check stock for each required part.
    If any part's quantity in the Service Warehouse is below the required amount,
    a Material Request is auto-created and assigned to the Parts Incharge.
    Source: PDF Section 9.2.
    """
    if not doc.get("service_items"):
        return

    items_to_request = []
    service_warehouse = frappe.db.get_single_value("Auto Dealer Settings", "service_warehouse") or "Stores - AD"

    for item in doc.service_items:
        if not item.item_code:
            continue
        actual_qty = flt(frappe.db.get_value(
            "Bin",
            {"item_code": item.item_code, "warehouse": service_warehouse},
            "actual_qty"
        ) or 0)
        required_qty = flt(item.qty or 1)
        if actual_qty < required_qty:
            items_to_request.append({
                "item_code": item.item_code,
                "qty": required_qty - actual_qty,
                "schedule_date": frappe.utils.today(),
                "warehouse": service_warehouse,
                "description": f"Auto-requested for Service Job Card {doc.name}",
            })

    if items_to_request:
        try:
            mr = frappe.get_doc({
                "doctype": "Material Request",
                "material_request_type": "Purchase",
                "transaction_date": frappe.utils.today(),
                "items": items_to_request,
            })
            mr.insert(ignore_permissions=True)
            frappe.msgprint(
                _("Material Request {0} auto-created for short-stock parts.").format(mr.name),
                alert=True, indicator="orange"
            )
        except Exception:
            frappe.log_error(
                title="Material Request Auto-Creation Failed",
                message=frappe.get_traceback()
            )


def _send_job_started_whatsapp(doc):
    """
    Send WhatsApp to customer that service has started.
    Template: service_started — name, job_card, vehicle, advisor (PDF Section 11.1).
    """
    try:
        settings = frappe.get_single("Auto Dealer Settings")
        if not settings.whatsapp_enabled:
            return
        from auto_dealer.api.whatsapp import send_whatsapp_message
        mobile = doc.get("mobile_no") or frappe.db.get_value("Customer", doc.customer, "mobile_no")
        if not mobile:
            return
        message = _(
            "Dear {0}, your vehicle {1} has been assigned to our workshop. "
            "Service Advisor: {2}. Job Card: {3}. "
            "We will notify you when your vehicle is ready."
        ).format(
            doc.customer_name or doc.customer,
            doc.vehicle,
            doc.service_advisor or "our team",
            doc.name,
        )
        send_whatsapp_message(mobile_no=mobile, message=message)
    except Exception:
        frappe.log_error(
            title="WhatsApp Service Started Notification Failed",
            message=frappe.get_traceback()
        )

File: auto_dealer/auto_dealer/public/js/service_job_card.js
frappe.ui.form.on("Service Job Card", {
	refresh: function (frm) {
		// Show "Send Ready Notification" button for completed jobs
		if (frm.doc.docstatus === 1 && frm.doc.status === "Completed" && frm.doc.mobile_no) {
			frm.add_custom_button(__("Send Ready SMS/WhatsApp"), function () {
				frappe.call({
					method: "auto_dealer.auto_dealer.api.whatsapp.send_whatsapp_message",
					args: {
						mobile_no: frm.doc.mobile_no,
						message: `Dear ${frm.doc.customer_name || frm.doc.customer}, your vehicle (VIN: ${frm.doc.vehicle}) is ready for pickup. Total: ₹${frm.doc.grand_total}. Thank you!`,
					},
					callback: function (r) {
						if (!r.exc) {
							frappe.show_alert({ message: __("Notification sent"), indicator: "green" });
						}
					},
				});
			}, __("Notify Customer"));
		}
	},

	// Auto-calculate line totals for parts
	service_items_on_form_rendered: function (frm) {
		frm.fields_dict.service_items.grid.on_row_add = function (row) {
			frappe.ui.form.on("Service Job Card Item", {
				qty: function (frm, cdt, cdn) {
					_calc_parts_row(frm, cdt, cdn);
				},
				rate: function (frm, cdt, cdn) {
					_calc_parts_row(frm, cdt, cdn);
				},
			});
		};
	},
});

function _calc_parts_row(frm, cdt, cdn) {
	const row = locals[cdt][cdn];
	const amount = (row.qty || 0) * (row.rate || 0);
	frappe.model.set_value(cdt, cdn, "amount", amount);
	frm.refresh_field("service_items");
}

frappe.ui.form.on("Service Job Card Item", {
	qty: function (frm, cdt, cdn) { _calc_parts_row(frm, cdt, cdn); },
	rate: function (frm, cdt, cdn) { _calc_parts_row(frm, cdt, cdn); },
});

frappe.ui.form.on("Service Job Card Labour", {
	hours: function (frm, cdt, cdn) {
		const row = locals[cdt][cdn];
		frappe.model.set_value(cdt, cdn, "amount", (row.hours || 0) * (row.rate_per_hour || 0));
		frm.refresh_field("labour_items");
	},
	rate_per_hour: function (frm, cdt, cdn) {
		const row = locals[cdt][cdn];
		frappe.model.set_value(cdt, cdn, "amount", (row.hours || 0) * (row.rate_per_hour || 0));
		frm.refresh_field("labour_items");
	},
});
